src/mgr/MDSPerfMetricTypes.h

```diff
@@ -0,0 +1,333 @@
+                     ceph::buffer::list::contiguous_appender& p) {
+    denc_varint(v.size(), p);
+    for (auto& i : v) {
+      denc(i, p);
+    }
+  }
+  static void decode(MDSPerformanceCounterDescriptors& v,
+                     ceph::buffer::ptr::const_iterator& p) {
+    unsigned num;
+    denc_varint(num, p);
+    v.clear();
+    v.reserve(num);
+    for (unsigned i=0; i < num; ++i) {
+      MDSPerformanceCounterDescriptor d;
+      denc(d, p);
+      if (d.is_supported()) {
+        v.push_back(std::move(d));
+      }
+    }
+  }
+};
+
+struct MDSPerfMetricLimit {
+  MDSPerformanceCounterDescriptor order_by;
+  uint64_t max_count;
+
+  MDSPerfMetricLimit() {
+  }
+  MDSPerfMetricLimit(const MDSPerformanceCounterDescriptor &order_by, uint64_t max_count)
+    : order_by(order_by), max_count(max_count) {
+  }
+
+  bool operator<(const MDSPerfMetricLimit &other) const {
+    if (order_by != other.order_by) {
+      return order_by < other.order_by;
+    }
+
+    return max_count < other.max_count;
+  }
+
+  DENC(MDSPerfMetricLimit, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.order_by, p);
+    denc(v.max_count, p);
+    DENC_FINISH(p);
+  }
+};
+WRITE_CLASS_DENC(MDSPerfMetricLimit)
+
+std::ostream &operator<<(std::ostream &os, const MDSPerfMetricLimit &limit);
+typedef std::set<MDSPerfMetricLimit> MDSPerfMetricLimits;
+
+struct MDSPerfMetricQuery {
+  MDSPerfMetricKeyDescriptor key_descriptor;
+  MDSPerformanceCounterDescriptors performance_counter_descriptors;
+
+  MDSPerfMetricQuery() {
+  }
+  MDSPerfMetricQuery(const MDSPerfMetricKeyDescriptor &key_descriptor,
+                     const MDSPerformanceCounterDescriptors &performance_counter_descriptors)
+    : key_descriptor(key_descriptor),
+      performance_counter_descriptors(performance_counter_descriptors)
+  {
+  }
+
+  bool operator<(const MDSPerfMetricQuery &other) const {
+    if (key_descriptor < other.key_descriptor) {
+      return true;
+    }
+    if (key_descriptor > other.key_descriptor) {
+      return false;
+    }
+    return performance_counter_descriptors < other.performance_counter_descriptors;
+  }
+
+  template <typename L>
+  bool get_key(L&& get_sub_key, MDSPerfMetricKey *key) const {
+    for (auto &sub_key_descriptor : key_descriptor) {
+      MDSPerfMetricSubKey sub_key;
+      if (!get_sub_key(sub_key_descriptor, &sub_key)) {
+        return false;
+      }
+      key->push_back(sub_key);
+    }
+    return true;
+  }
+
+  void get_performance_counter_descriptors(MDSPerformanceCounterDescriptors *descriptors) const {
+    *descriptors = performance_counter_descriptors;
+  }
+
+  template <typename L>
+  void update_counters(L &&update_counter, PerformanceCounters *counters) const {
+    auto it = counters->begin();
+    for (auto &descriptor : performance_counter_descriptors) {
+      // TODO: optimize
+      if (it == counters->end()) {
+        counters->push_back(PerformanceCounter());
+        it = std::prev(counters->end());
```

Comment:
```c++
std::advance(it, -1);
```

Reply:
probably when we templatize this (and the respective OSD class), all this would be cleaned up.

Reply:
sorry, i was wrong. as `push_back()` changes the size of `counters`, `counters` could be reallocated, so `it` is not guaranteed to be a validate iterator after the vector is resized.